src/Persistence_matrix/include/gudhi/Persistence_matrix/columns/heap_column.h

```diff
@@ -0,0 +1,1082 @@
+                "For boundary columns, the range has to be a column of same type to help ensure the validity of the "
+                "base element.");  // could be removed, if we give the responsability to the user.
+  static_assert((!Master_matrix::isNonBasic || Master_matrix::Option_list::is_of_boundary_type),
+                "For chain columns, the given column cannot be constant.");
+
+  if constexpr (Master_matrix::Option_list::is_z2) {
+    if (val) {
+      _add(column);
+    }
+  } else {
+    _multiply_and_add(column, val);
+  }
+
+  return *this;
+}
+
+template <class Master_matrix>
+inline Heap_column<Master_matrix>& Heap_column<Master_matrix>::multiply_and_add(
+    Heap_column& column, const Field_element_type& val) 
+{
+  if constexpr (Master_matrix::isNonBasic && !Master_matrix::Option_list::is_of_boundary_type) {
+    // assumes that the addition never zeros out this column.
+    if constexpr (Master_matrix::Option_list::is_z2) {
+      if (val) {
+        if (_add(column)) {
+          chain_opt::swap_pivots(column);
+          dim_opt::swap_dimension(column);
+        }
+      }
+    } else {
+      if (_multiply_and_add(column, val)) {
+        chain_opt::swap_pivots(column);
+        dim_opt::swap_dimension(column);
+      }
+    }
+  } else {
+    if constexpr (Master_matrix::Option_list::is_z2) {
+      if (val) {
+        _add(column);
+      }
+    } else {
+      _multiply_and_add(column, val);
+    }
+  }
+
+  return *this;
+}
+
+template <class Master_matrix>
+inline Heap_column<Master_matrix>& Heap_column<Master_matrix>::operator=(
+    const Heap_column& other) 
+{
+  static_assert(!Master_matrix::Option_list::has_row_access, "= assignement not enabled with row access option.");
+
+  dim_opt::operator=(other);
+  chain_opt::operator=(other);
+
+  while (column_.size() > other.column_.size()) {
+    if (column_.back() != nullptr) cellPool_->destroy(column_.back());
+    column_.pop_back();
+  }
+
+  column_.resize(other.column_.size(), nullptr);
+  index i = 0;
+  for (const Cell* cell : other.column_) {
+    if (column_[i] != nullptr) {
+      cellPool_->destroy(column_[i]);
+    }
+    if constexpr (Master_matrix::Option_list::is_z2) {
+      column_[i++] = other.cellPool_->construct(cell->get_row_index());
+    } else {
+      column_[i] = other.cellPool_->construct(cell->get_row_index());
+      column_[i++]->set_element(cell->get_element());
+    }
+  }
+  insertsSinceLastPrune_ = other.insertsSinceLastPrune_;
+  operators_ = other.operators_;
+  cellPool_ = other.cellPool_;
+
+  return *this;
+}
+
+template <class Master_matrix>
+inline void Heap_column<Master_matrix>::_prune() 
+{
+  if (insertsSinceLastPrune_ == 0) return;
+
+  Column_type tempCol;
+  Cell* pivot = _pop_pivot();
+  while (pivot != nullptr) {
+    tempCol.push_back(pivot);
+    pivot = _pop_pivot();
+  }
+  column_.swap(tempCol);
+  std::make_heap(column_.begin(), column_.end(), cellPointerComp_);
```

Comment:
Isn't column_ already sorted at this point, and hence already a heap?

Reply:
I don't think so. If I remember well the heap is sorted as a type of tree and therefore does not corresponds to a vector ordered by row indices.

Reply:
A min-heap is a tree where each node has a value smaller than its children. The vector is interpreted as a binary tree where node i has children 2i+1 and 2i+2. It seems to me that if the vector is sorted, then it also satisfies the (weaker) heap condition, no? (well, maybe the order needs to be reversed, I didn't check, but that's still easier than make_heap)

Reply:
Ok, for some reasons I thought that it was more complicated than that. But if the only condition is that the parent has to be smaller than the children, then the ordered vector should indeed work. I removed the line and it seems to work.